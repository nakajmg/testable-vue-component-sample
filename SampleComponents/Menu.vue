<template>
  <nav>
    <div v-show="menuOpened">
      <router-link to="/">Root</router-link>
      <router-link to="/foo">Foo</router-link>
      <router-link to="/bar">Bar</router-link>
    </div>
    <button @click="toggleMenu">
      Open Menu
    </button>
  </nav>
</template>

<script>
export default {
  name: "Menu",
  data() {
    return {
      menuOpened: false,
    }
  },
  methods: {
    toggleMenu() {
      this.menuOpened = !this.menuOpened
    },
  },
}
</script>

