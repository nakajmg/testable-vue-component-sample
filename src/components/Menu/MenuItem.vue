<template>
  <div class="MenuItem">
    <span class="MenuItem_Label"
      @click="clickMenuItem"
    >
      {{label}}
    </span>
  </div>
</template>

<script>
export default {
  name: "MenuItem",
  props: {
    label: String,
    name: String,
  },
  methods: {
    clickMenuItem() {
      this.$emit("clickMenuItem", { name: this.name })
    },
  },
}
</script>

<style lang="scss">
.MenuItem {
  &_Label {
    cursor: pointer;
    &:hover {
      text-decoration: underline;
    }
  }
}
</style>
