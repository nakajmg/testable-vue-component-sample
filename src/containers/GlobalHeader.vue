<template>
  <header class="Header">
    <Logo class="Header_Logo" />
    <span class="Header_SiteTitle"
      @click="navigateRoot"
    >
      <SiteTitle :title="siteTitle"/>
    </span>
    <Menu class="Header_Menu"
      :items="menuItems"
      @clickMenuItem="onClickMenuItem"
    />
  </header>
</template>

<script>
import { mapState } from "vuex"
import Logo from "../basics/Logo.vue"
import SiteTitle from "../basics/SiteTitle.vue"
import Menu from "../components/Menu.vue"
export default {
  name: "Header",
  components: {
    Logo,
    SiteTitle,
    Menu,
  },
  computed: {
    ...mapState(["siteTitle", "menuItems"]),
  },
  methods: {
    onClickMenuItem({ name }) {
      this.$emit("navigate", { name })
    },
    navigateRoot() {
      this.$emit("navigate", { name: "root" })
    },
  },
}
</script>

<style lang="scss">
.Header {
  padding: 0.5em 1em;
  background-color: #3eaf7c;
  display: flex;
  align-items: center;
  &_Logo {
    width: 35px;
  }
  &_SiteTitle {
    color: #fff;
    font-size: 1.75em;
    margin-left: 0.5em;
    cursor: pointer;
  }
  &_Menu {
    color: #fff;
    font-size: 1.25em;
    margin-left: auto;
  }
}
</style>
