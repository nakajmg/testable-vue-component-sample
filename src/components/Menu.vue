<template>
  <div class="Menu">
    <MenuItem class="Menu_Item"
      v-for="item in items"
      :key="item.label"
      v-bind="item"
      @clickMenuItem="onClickMenuItem"
    />
  </div>
</template>

<script>
import MenuItem from "../basics/MenuItem.vue"
export default {
  name: "Menu",
  components: {
    MenuItem,
  },
  props: {
    items: Array,
  },
  methods: {
    onClickMenuItem({ name }) {
      this.$emit("clickMenuItem", { name })
    },
  },
}
</script>

<style lang="scss">
.Menu {
  display: flex;
  align-items: center;
  &_Item + &_Item {
    margin-left: 1em;
  }
}
</style>
