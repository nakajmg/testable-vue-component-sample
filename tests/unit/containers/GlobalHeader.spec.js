import { shallowMount, mount, createLocalVue } from "@vue/test-utils"
import GlobalHeader from "@/containers/GlobalHeader.vue"
import Menu from "@/components/Menu.vue"
import Vuex from "vuex"
import menuItems from "../../_mockData/menuItems.json"
const localVue = createLocalVue()

localVue.use(Vuex)
describe("GlobalHeader.vue", () => {
  let store

  beforeAll(() => {
    store = new Vuex.Store({
      state: {
        siteTitle: "test site title",
        menuItems,
      },
    })
  })

  describe("methods", () => {
    it("clickMenuItem", () => {
      const wrapper = shallowMount(GlobalHeader, { store, localVue })
      wrapper.vm.onClickMenuItem(menuItems[0])
      expect(wrapper.emitted("navigate")).toBeTruthy()
      expect(wrapper.emitted("navigate")[0][0]).toEqual({
        name: menuItems[0].name,
      })
    })
    it("navigateRoot", () => {
      const wrapper = shallowMount(GlobalHeader, { store, localVue })
      wrapper.vm.navigateRoot()
      expect(wrapper.emitted("navigate")).toBeTruthy()
      expect(wrapper.emitted("navigate")[0][0]).toEqual({
        name: "root",
      })
    })
  })

  describe("template", () => {
    it("@click=navigateRoot", () => {
      const mock = jest.fn()
      const wrapper = shallowMount(GlobalHeader, { store, localVue })
      wrapper.setMethods({
        navigateRoot: mock,
      })
      wrapper.find(".Header_SiteTitle").trigger("click")
      expect(mock).toHaveBeenCalled()
    })
    it("@clickMenuItem=onClickMenuItem", () => {
      const mock = jest.fn()
      const wrapper = shallowMount(GlobalHeader, { store, localVue })
      wrapper.setMethods({
        onClickMenuItem: mock,
      })
      const menu = wrapper.find(Menu)
      menu.vm.$emit("clickMenuItem")
      expect(mock).toHaveBeenCalled()
    })
    it("snapshot", () => {
      const wrapper = mount(GlobalHeader, { store, localVue })
      expect(wrapper.vm.$el).toMatchSnapshot()
    })
  })
})
