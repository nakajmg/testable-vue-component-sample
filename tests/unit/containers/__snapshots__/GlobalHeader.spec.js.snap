// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalHeader.vue template snapshot 1`] = `
<header
  class="GlobalHeader"
>
  <div
    class="Logo GlobalHeader_Logo"
  >
    <img
      alt="Vue logo"
      class="Logo_Img"
      src="../assets/logo.png"
    />
  </div>
   
  <span
    class="GlobalHeader_SiteTitle"
  >
    <div
      class="SiteTitle"
    >
      
  test site title

    </div>
  </span>
   
  <div
    class="Menu GlobalHeader_Menu"
  >
    <div
      class="MenuItem Menu_Item"
    >
      <span
        class="MenuItem_Label"
      >
        
    About
  
      </span>
    </div>
    <div
      class="MenuItem Menu_Item"
    >
      <span
        class="MenuItem_Label"
      >
        
    Todos
  
      </span>
    </div>
  </div>
</header>
`;
